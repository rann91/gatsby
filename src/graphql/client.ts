import { ApolloClient, InMemoryCache, HttpLink } from '@apollo/client'
import fetch from 'cross-fetch'

const client = new ApolloClient({
  link: new HttpLink({
    uri: `${process.env.GATSBY_API_URL}/graphql`,
    fetch
  }),
  cache: new InMemoryCache()
})

export default client
