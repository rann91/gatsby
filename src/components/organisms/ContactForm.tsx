/* @jsx jsx */
import { gql, useMutation } from '@apollo/client'
import { useForm } from 'react-hook-form'
import ReactMarkdown from 'react-markdown'
import {
  Box,
  Button,
  Container,
  Flex,
  Heading,
  Input,
  jsx,
  Label,
  Text,
  Textarea
} from 'theme-ui'
import { column } from '../../gatsby-plugin-theme-ui'
import {
  Strapi_ComponentContentContactForm,
  Strapi_ComponentSharedContactDetail
} from '../../typings/graphql'
import Section from '../atoms/Section'

interface Input {
  name: string
  email: string
  message: string
}

interface Data {
  contact: {
    success: boolean
  }
}

interface Variables {
  data: Input
}

const ContactForm = ({
  title,
  description,
  contactDetail
}: Strapi_ComponentContentContactForm) => {
  const [sendForm, { data, loading, error }] = useMutation<Data, Variables>(gql`
    mutation SendForm($data: ContactInput!) {
      contact(data: $data) {
        success
      }
    }
  `)
  const { register, handleSubmit } = useForm<Input>()
  const onSubmit = async (data: Input) => {
    await sendForm({ variables: { data } })
  }

  return (
    <Section hasTitle={!!title}>
      <Container>
        <Flex sx={{ flexWrap: 'wrap' }}>
          <Box px={5} sx={{ width: [column(12), null, null, column(8)] }}>
            <Box
              p={[5, null, null, 6]}
              sx={{
                borderWidth: 1,
                borderStyle: 'solid',
                borderColor: 'muted'
              }}>
              {title && (
                <Heading as="h2" variant="h2">
                  {title}
                </Heading>
              )}
              {description && (
                <Text
                  variant="p"
                  pb={5}
                  sx={{
                    '* + &': {
                      pt: 3
                    }
                  }}>
                  {description}
                </Text>
              )}
              {error || data?.contact.success === false ? (
                <Text variant="p">Something went wrong.</Text>
              ) : data?.contact.success ? (
                <Text variant="p">Message successfully sent!</Text>
              ) : (
                <form onSubmit={handleSubmit(onSubmit)}>
                  <Box pb={5}>
                    <Label htmlFor="name">Name</Label>
                    <Input
                      ref={register}
                      id="name"
                      name="name"
                      type="text"
                      required={true}
                      placeholder="Enter your name"
                    />
                  </Box>
                  <Box pb={5}>
                    <Label htmlFor="email">Email address</Label>
                    <Input
                      ref={register}
                      id="email"
                      name="email"
                      type="email"
                      required={true}
                      placeholder="Enter you email"
                    />
                  </Box>
                  <Box pb={5}>
                    <Label htmlFor="message">Message</Label>
                    <Textarea
                      ref={register}
                      id="message"
                      name="message"
                      placeholder="Enter your message"
                      rows={5}
                      required={true}
                      sx={{ resize: 'none' }}
                    />
                  </Box>
                  <Button
                    type="submit"
                    disabled={loading}
                    sx={{ display: 'block', width: '100%' }}>
                    Submit
                  </Button>
                </form>
              )}
            </Box>
          </Box>
          <Box
            px={5}
            sx={{
              width: [column(12), null, null, column(4)],
              '* + &': {
                pt: [6, null, null, 0]
              }
            }}>
            {((contactDetail ||
              []) as Strapi_ComponentSharedContactDetail[]).map(
              ({ id, title, content }) => (
                <Box key={id} sx={{ '&:last-child': { mb: -1 * 5 } }}>
                  {title && (
                    <Heading
                      as="h3"
                      variant="label"
                      pb={2}
                      sx={{ color: 'text' }}>
                      {title}
                    </Heading>
                  )}
                  {content && (
                    <ReactMarkdown
                      source={content}
                      sx={{
                        p: {
                          variant: 'text.p',
                          m: 0,
                          pb: 5,
                          whiteSpace: 'pre-wrap'
                        },
                        a: {
                          color: 'inherit'
                        }
                      }}
                    />
                  )}
                </Box>
              )
            )}
          </Box>
        </Flex>
      </Container>
    </Section>
  )
}

export default ContactForm
