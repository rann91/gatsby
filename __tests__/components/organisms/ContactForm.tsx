import { ApolloProvider } from '@apollo/client'
import { render } from '@testing-library/react'
import React from 'react'
import ContactForm from '../../../src/components/organisms/ContactForm'
import client from '../../../src/graphql/client'
import component from '../../fixtures/component-contact-form.json'

describe('ContactForm', () => {
  it('renders correctly', () => {
    const { container } = render(
      <ApolloProvider client={client}>
        <ContactForm {...component} />
      </ApolloProvider>
    )

    expect(container).toMatchSnapshot()
  })
})
