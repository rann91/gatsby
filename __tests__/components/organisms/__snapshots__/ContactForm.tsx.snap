// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm renders correctly 1`] = `
<div>
  <section
    class="css-1ptauik"
    data-testid="section"
  >
    <div
      class="css-1yrmf69"
    >
      <div
        class="css-kj8tif-ContactForm"
      >
        <div
          class="css-6lpdcm-ContactForm"
        >
          <div
            class="css-e4xjql-ContactForm"
          >
            <h2
              class="css-1wf7gym"
            >
              Contact us
            </h2>
            <div
              class="css-u09dmq-ContactForm"
            >
              Even the all-powerful Pointing has no control about the blind texts it is an almost unorthographic life One day however a small line of blind text by the name of Lorem Ipsum decided to leave for the far World of Grammar.
            </div>
            <form>
              <div
                class="css-1pd4dm3"
              >
                <label
                  class="css-knfhfe"
                  for="name"
                >
                  Name
                </label>
                <input
                  class="css-iyqgnx"
                  id="name"
                  name="name"
                  placeholder="Enter your name"
                  required=""
                  type="text"
                />
              </div>
              <div
                class="css-1pd4dm3"
              >
                <label
                  class="css-knfhfe"
                  for="email"
                >
                  Email address
                </label>
                <input
                  class="css-iyqgnx"
                  id="email"
                  name="email"
                  placeholder="Enter you email"
                  required=""
                  type="email"
                />
              </div>
              <div
                class="css-1pd4dm3"
              >
                <label
                  class="css-knfhfe"
                  for="message"
                >
                  Message
                </label>
                <textarea
                  class="css-g7i2cv-ContactForm"
                  id="message"
                  name="message"
                  placeholder="Enter your message"
                  required=""
                  rows="5"
                />
              </div>
              <button
                class="css-1guur8s-ContactForm"
                type="submit"
              >
                Submit
              </button>
            </form>
          </div>
        </div>
        <div
          class="css-180oxhi-ContactForm"
        >
          <div
            class="css-vnrgx5"
          >
            <h3
              class="css-19rfdr3"
            >
              Our offices
            </h3>
            <div
              class="css-qilbtt"
            >
              <p>
                Even the all-powerful Pointing has no control about the blind texts it is an almost unorthographic.
              </p>
            </div>
          </div>
          <div
            class="css-vnrgx5"
          >
            <h3
              class="css-19rfdr3"
            >
              Our office
            </h3>
            <div
              class="css-qilbtt"
            >
              <p>
                908 Broadway,
San Francisco,
CA 94133
              </p>
            </div>
          </div>
          <div
            class="css-vnrgx5"
          >
            <h3
              class="css-19rfdr3"
            >
              Working hours
            </h3>
            <div
              class="css-qilbtt"
            >
              <p>
                9AM - 3PM, Mon to Fri
              </p>
            </div>
          </div>
          <div
            class="css-vnrgx5"
          >
            <h3
              class="css-19rfdr3"
            >
              Contact
            </h3>
            <div
              class="css-qilbtt"
            >
              <p>
                <a
                  href="mailto:contact@business.com"
                >
                  contact@business.com
                </a>
                
01 (650) 658 6822
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
